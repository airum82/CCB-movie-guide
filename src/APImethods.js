
export const getMoviesByCategory = (category, page = 1) => {
  return fetch(`https://api.themoviedb.org/3/movie/${category}?api_key=${process.env.APIkey}&language=en-US&page=${page}`)
    .then(response => response.json())
    .catch(err => console.log(err));
}

export const getMovieDetails = movieId => {
  return fetch(`https://api.themoviedb.org/3/movie/${movieId}?api_key=${process.env.APIkey}&language=en-US`)
    .then(response => response.json())
    .catch(err => console.log(err));
}