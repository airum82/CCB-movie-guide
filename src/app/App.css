:root {
  --color1: rgba(0, 78, 100, 1);
  --color2: rgba(0, 165, 207, 1);
  --color3: rgb(179, 231, 202);
  --color4: rgba(37, 161, 142, 1);
  --color5: rgba(122, 229, 130, 1);
}

body {
  background-color: var(--color3);
}

.title {
  text-align: center;
}
