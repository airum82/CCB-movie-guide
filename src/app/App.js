import React, { Component } from 'react';
import Header from '../header/Header';
import MovieContainer from '../movies-container/MoviesContainer';
import * as API from '../APImethods';
import { Route, withRouter } from 'react-router-dom';
import 'normalize.css';
import './App.css';

class App extends Component {
  constructor() {
    super()
    this.state = {
      now_playing: [],
      popular: [],
      top_rated: [],
      searchTerms: '',
      searchResults: [],
      movie: []
    }
    this.grabSearchTerms = this.grabSearchTerms.bind(this);
    this.getNewCategory = this.getNewCategory.bind(this);
    this.searchMovies = this.searchMovies.bind(this);
    this.viewMovie = this.viewMovie.bind(this);
    this.formatReleaseDate = this.formatReleaseDate.bind(this);
  }

  grabSearchTerms(e) {
    this.setState({
      searchTerms: (e.target.value).toLowerCase()
    })
  }

  searchMovies() {
    event.preventDefault();
    ['popular', 'top_rated'].forEach((category, i, cats) => {
      if (!this.state[category].length) {
        API.getMoviesByCategory(category)
          .then(movies => this.setState({ [category]: movies.results }))
          .then(() => {
            if(i === cats.length - 1) {
              this.createResults();
            }
          }).then(() => this.props.history.push('/searchResults'))
      } else if(i === cats.length - 1) {
          this.createResults();
      }
    })
    this.props.history.push('/searchResults')
  }

  createResults() {
    const { now_playing, popular, top_rated } = this.state;
    const results = [...now_playing, ...popular, ...top_rated].reduce((results, movie) => {
      if (movie.title.toLowerCase().includes(this.state.searchTerms)) {
        const duplicate = results.find(result =>
          result.title.toLowerCase() === movie.title.toLowerCase())
        if (!duplicate) results.push(movie);
      }
      return results;
    }, [])
    this.setState({ 
      searchResults: results,
      searchTerms: '' 
    })
  }

  getNewCategory(e) {
    const category = e.target.id;
    if(!this.state[category].length) {
      API.getMoviesByCategory(category)
        .then(movies => this.setState({ [category]: movies.results }))
    }
  }

  viewMovie(id) {
    API.getMovieDetails(id)
      .then(movie => {
        this.setState({ movie: [movie] });
        return movie;
      })
      .then(movie => {
        if(this.props.location.pathname !== `/movie/${movie.id}`) {
          this.props.history.push(`/movie/${movie.id}`)
        }
      })
  }

  formatReleaseDate(date) {
    const dateArray = date.split('-');
    return `${dateArray[1]}-${dateArray[2]}-${dateArray[0]}`
  }

  componentDidMount() {
    API.getMoviesByCategory('now_playing')
      .then(movies => this.setState({ now_playing: movies.results }))
      .then(() => this.props.history.push('/now_playing'))
  }

  render() {
    return (
      <div>
        <h1 
          className="title"
          onClick={() => this.props.history.push('/now_playing')}
        >Ultimate Movie Guide</h1>
        <Header 
          grabSearchTerms={this.grabSearchTerms} 
          getNewCategory={this.getNewCategory}
          searchMovies={this.searchMovies}
          location={this.props.location}
          searchTerms={this.state.searchTerms}
        />
        <Route path='/:category' render={({ match }) => {
          const category = match.params.category;
          return (
            <MovieContainer 
              movies={this.state[category]}
              viewMovie={this.viewMovie}
              location={this.props.location}
              formatReleaseDate={this.formatReleaseDate}
            />
          )
        }} />
      </div>
    )
  }
}

export default withRouter(App);