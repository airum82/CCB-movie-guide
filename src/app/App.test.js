import React from 'react';
import { App } from './App';
import { shallow } from 'enzyme';
import * as API from '../APImethods';

describe('App', () => {
  let wrapper;
  const mockProps = {
    history: {
      push: jest.fn()
    }
  }
  beforeEach(() => {
    wrapper = shallow(<App {...mockProps} />, {
      disableLifecycleMethods: true
    })
  })

  it('should match snapshot upon render', () => {
    expect(wrapper).toMatchSnapshot();
  })

  it('grab search terms should set state with lower case string', () => {
    const mockEvent = {
      target: {
        value: 'Rob'
      }
    };
    const expectedResult = mockEvent.target.value.toLowerCase();
    expect(wrapper.state('searchTerms')).toEqual('');
    wrapper.instance().grabSearchTerms(mockEvent);
    expect(wrapper.state('searchTerms')).toEqual(expectedResult);
  })

  it('searchMovies should call API.getMoviesByCategory and createResults', () => {
    API.getMoviesByCategory = jest.fn().mockImplementation(() => Promise.resolve({
      json: () => Promise.resolve({})
    }));
    wrapper.instance().createResults = jest.fn();
    const mockEvent = {
      preventDefault: jest.fn()
    }
    wrapper.instance().searchMovies(mockEvent);
    expect(API.getMoviesByCategory).toHaveBeenCalled();
  })

  it('createResults should setState with results that match search terms', () => {
    wrapper.instance().setState({
      now_playing: [{ title: 'Robin' }, { title: 'Batman'}, { title: 'Robinson'}],
      popular: [{ title: 'Parent Trap'}, { title: 'Robbery'}]
    });
    const mockEvent = {
      target: {
        value: 'Rob'
      }
    };
    const expectedResult = [
      { title: 'Robin' }, 
      { title: 'Robinson' }, 
      { title: 'Robbery' }
    ]
    wrapper.instance().grabSearchTerms(mockEvent);
    wrapper.instance().createResults();
    expect(wrapper.state('searchResults')).toEqual(expectedResult);
  })

  //test get new category later
  it('get new category should call API.getMoviesByCategory', () => {
    API.getMoviesByCategory = jest.fn().mockImplementation(() => Promise.resolve({
      json: () => Promise.resolve({})
    }));
    const mockEvent = {
      target: {
        id: 'popular'
      }
    }
    wrapper.instance().getNewCategory(mockEvent)
    expect(API.getMoviesByCategory).toHaveBeenCalledWith(mockEvent.target.id)
  })
  
  //test viewMovie later

  it('formatReleaseDate should format the imdb date to american standards', () => {
    const rawDate = '2019-01-15';
    const expectedResult = '01-15-2019';
    const result = wrapper.instance().formatReleaseDate(rawDate);
    expect(result).toEqual(expectedResult);
  })

})
