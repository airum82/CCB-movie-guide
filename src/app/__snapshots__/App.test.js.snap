// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot upon render 1`] = `
<div>
  <h1
    className="title"
    onClick={[Function]}
  >
    Ultimate Movie Guide
  </h1>
  <Header
    getNewCategory={[Function]}
    grabSearchTerms={[Function]}
    location={
      Object {
        "pathname": "/",
      }
    }
    searchMovies={[Function]}
    searchTerms=""
  />
  <div
    className="loader"
  >
    <Loader
      color="#71238"
      height="100"
      type="Grid"
      width="100"
    />
  </div>
  <button
    className="load-more"
    onClick={[Function]}
  >
    Load More
  </button>
</div>
`;
